File: examples/05-errors-three-ways/go/main.go
// copyx SRC DST — file copier that establishes the book's error taxonomy.
//
//	exit 0 — success; "copied <N> bytes" on stdout
//	exit 2 — source-side failure (open/read SRC); "copyx: <reason>" on stderr
//	exit 3 — destination-side failure (open/write/close DST); same shape
//
// EINTR on read(2)/write(2) is retried (the syscall was interrupted before
// completing; reissuing it is the only correct policy), and short writes are
// resumed until the whole buffer is on its way.
package main

import (
	"errors"
	"fmt"
	"os"
	"unicode"

	"golang.org/x/sys/unix"
)

// opError pins a failure to a syscall phase so main can map it onto the
// taxonomy: source-side failures exit 2, destination-side failures exit 3.
type opError struct {
	op   string // "read", "write", "close"; "" means open(2)
	path string
	err  unix.Errno // reachable via errors.Is / errors.As through Unwrap
	exit int
}

func (e *opError) Error() string {
	if e.op == "" {
		return fmt.Sprintf("%s: %s", e.path, strerror(e.err))
	}
	return fmt.Sprintf("%s %s: %s", e.op, e.path, strerror(e.err))
}

func (e *opError) Unwrap() error { return e.err }

// newOpError wraps whatever the unix package returned; the raw errno is
// recovered with errors.As so the chain stays inspectable.
func newOpError(op, path string, err error, exit int) *opError {
	var errno unix.Errno
	if !errors.As(err, &errno) {
		errno = unix.EIO
	}
	return &opError{op: op, path: path, err: errno, exit: exit}
}

// strerror renders an errno the way glibc's strerror(3) does — first letter
// capitalized — so all three implementations print identical reason text.
func strerror(errno unix.Errno) string {
	r := []rune(errno.Error())
	if len(r) > 0 {
		r[0] = unicode.ToUpper(r[0])
	}
	return string(r)
}

// readSome issues read(2) once, retrying EINTR: interrupted means nothing
// was consumed, so reissue the call.
func readSome(fd int, buf []byte) (int, error) {
	for {
		n, err := unix.Read(fd, buf)
		if errors.Is(err, unix.EINTR) {
			continue // interrupted before transferring anything: retry
		}
		return n, err
	}
}

// writeAll pushes the whole buffer through write(2): EINTR restarts the
// call, a short write resumes from where the kernel stopped.
func writeAll(fd int, buf []byte) error {
	for len(buf) > 0 {
		n, err := unix.Write(fd, buf)
		if errors.Is(err, unix.EINTR) {
			continue // interrupted: reissue the same span
		}
		if err != nil {
			return err
		}
		buf = buf[n:]
	}
	return nil
}

func copyFile(srcPath, dstPath string) (int64, error) {
	src, err := unix.Open(srcPath, unix.O_RDONLY|unix.O_CLOEXEC, 0)
	if err != nil {
		return 0, newOpError("", srcPath, fmt.Errorf("open %s: %w", srcPath, err), 2)
	}
	defer unix.Close(src) // read side: nothing actionable in the close result

	dst, err := unix.Open(dstPath, unix.O_WRONLY|unix.O_CREAT|unix.O_TRUNC|unix.O_CLOEXEC, 0o644)
	if err != nil {
		return 0, newOpError("", dstPath, fmt.Errorf("open %s: %w", dstPath, err), 3)
	}

	var total int64
	buf := make([]byte, 64*1024)
	for {
		n, err := readSome(src, buf)
		if err != nil {
			unix.Close(dst)
			return 0, newOpError("read", srcPath, err, 2)
		}
		if n == 0 {
			break // EOF
		}
		if err := writeAll(dst, buf[:n]); err != nil {
			unix.Close(dst)
			return 0, newOpError("write", dstPath, err, 3)
		}
		total += int64(n)
	}

	// The write side must observe close(2): deferred IO errors land here.
	if err := unix.Close(dst); err != nil {
		return 0, newOpError("close", dstPath, err, 3)
	}
	return total, nil
}

func main() {
	if len(os.Args) != 3 {
		fmt.Fprintln(os.Stderr, "usage: copyx SRC DST")
		os.Exit(2)
	}
	n, err := copyFile(os.Args[1], os.Args[2])
	if err != nil {
		fmt.Fprintf(os.Stderr, "copyx: %s\n", err)
		var op *opError
		if errors.As(err, &op) {
			os.Exit(op.exit)
		}
		os.Exit(3)
	}
	fmt.Printf("copied %d bytes\n", n)
}

File: examples/05-errors-three-ways/rust/src/main.rs
//! copyx SRC DST — file copier that establishes the book's error taxonomy.
//!
//!   exit 0 — success; "copied <N> bytes" on stdout
//!   exit 2 — source-side failure (open/read SRC); "copyx: <reason>" on stderr
//!   exit 3 — destination-side failure (open/write/close DST); same shape
//!
//! EINTR on read(2)/write(2) is retried (the syscall was interrupted before
//! completing; reissuing it is the only correct policy), and short writes are
//! resumed until the whole buffer is on its way.

use std::os::fd::{AsFd, IntoRawFd, OwnedFd};
use std::process::ExitCode;

use rustix::fs::{Mode, OFlags, open};
use rustix::io::{Errno, read, try_close, write};
use thiserror::Error;

/// Each variant pins a failure to a syscall phase so `main` can map it onto
/// the taxonomy: source-side failures exit 2, destination-side failures 3.
#[derive(Debug, Error)]
enum CopyError {
    #[error("{path}: {reason}")]
    SrcOpen { path: String, reason: String },
    #[error("read {path}: {reason}")]
    Read { path: String, reason: String },
    #[error("{path}: {reason}")]
    DstOpen { path: String, reason: String },
    #[error("write {path}: {reason}")]
    Write { path: String, reason: String },
    #[error("close {path}: {reason}")]
    Close { path: String, reason: String },
}

impl CopyError {
    fn exit_code(&self) -> u8 {
        match self {
            CopyError::SrcOpen { .. } | CopyError::Read { .. } => 2,
            CopyError::DstOpen { .. } | CopyError::Write { .. } | CopyError::Close { .. } => 3,
        }
    }
}

/// Render an errno the way glibc's strerror(3) does, so all three
/// implementations print identical reason text. `Errno`'s own Display goes
/// through `std::io::Error`, which appends " (os error N)" — strip it.
fn strerror(errno: Errno) -> String {
    let s = std::io::Error::from(errno).to_string();
    match s.rfind(" (os error ") {
        Some(idx) => s[..idx].to_owned(),
        None => s,
    }
}

/// read(2) once, retrying EINTR: interrupted means nothing was consumed.
fn read_some(fd: impl AsFd, buf: &mut [u8]) -> Result<usize, Errno> {
    loop {
        match read(&fd, &mut *buf) {
            Err(Errno::INTR) => continue, // interrupted before transferring anything: retry
            other => return other,
        }
    }
}

/// write(2) until the whole slice is written: EINTR restarts the call, a
/// short write resumes from where the kernel stopped.
fn write_all(fd: impl AsFd, mut buf: &[u8]) -> Result<(), Errno> {
    while !buf.is_empty() {
        match write(&fd, buf) {
            Ok(n) => buf = &buf[n..],
            Err(Errno::INTR) => continue, // interrupted: reissue the same span
            Err(e) => return Err(e),
        }
    }
    Ok(())
}

/// close(2) with the result observed: on the write side this is where
/// deferred IO errors surface. `into_raw_fd` transfers ownership out of the
/// `OwnedFd`, so the fd is closed exactly once.
fn close_checked(fd: OwnedFd) -> Result<(), Errno> {
    unsafe { try_close(fd.into_raw_fd()) }
}

fn copy_file(src_path: &str, dst_path: &str) -> Result<u64, CopyError> {
    let src = open(src_path, OFlags::RDONLY | OFlags::CLOEXEC, Mode::empty()).map_err(|e| {
        CopyError::SrcOpen { path: src_path.to_owned(), reason: strerror(e) }
    })?; // OwnedFd: dropped (and closed) on every path below

    let dst = open(
        dst_path,
        OFlags::WRONLY | OFlags::CREATE | OFlags::TRUNC | OFlags::CLOEXEC,
        Mode::from_bits_truncate(0o644),
    )
    .map_err(|e| CopyError::DstOpen { path: dst_path.to_owned(), reason: strerror(e) })?;

    let mut total: u64 = 0;
    let mut buf = vec![0u8; 64 * 1024];
    loop {
        let n = read_some(&src, &mut buf).map_err(|e| CopyError::Read {
            path: src_path.to_owned(),
            reason: strerror(e),
        })?;
        if n == 0 {
            break; // EOF
        }
        write_all(&dst, &buf[..n]).map_err(|e| CopyError::Write {
            path: dst_path.to_owned(),
            reason: strerror(e),
        })?;
        total += n as u64;
    }

    close_checked(dst).map_err(|e| CopyError::Close {
        path: dst_path.to_owned(),
        reason: strerror(e),
    })?;
    Ok(total)
}

fn main() -> ExitCode {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 3 {
        eprintln!("usage: copyx SRC DST");
        return ExitCode::from(2);
    }
    match copy_file(&args[1], &args[2]) {
        Ok(n) => {
            println!("copied {n} bytes");
            ExitCode::SUCCESS
        }
        Err(e) => {
            eprintln!("copyx: {e}");
            ExitCode::from(e.exit_code())
        }
    }
}
